cmd/juju/cloud/add_test.go

```diff
@@ -7,172 +7,256 @@ import (
 	"bytes"
 	"io"
 	"io/ioutil"
-	"path/filepath"
 	"strings"
 
 	"github.com/juju/cmd"
+	jujutesting "github.com/juju/testing"
 	jc "github.com/juju/testing/checkers"
 	"github.com/pkg/errors"
 	gc "gopkg.in/check.v1"
 
+	cloudfile "github.com/juju/juju/cloud"
 	"github.com/juju/juju/cmd/juju/cloud"
-	"github.com/juju/juju/juju/osenv"
 	"github.com/juju/juju/testing"
 )
 
 type addSuite struct {
-	testing.FakeJujuXDGDataHomeSuite
+	jujutesting.IsolationSuite
 }
 
 var _ = gc.Suite(&addSuite{})
 
+func newFakeCloudQueryer() *fakeCloudQueryer {
+	return &fakeCloudQueryer{CallMocker: jujutesting.NewCallMocker()}
+}
+
+type fakeCloudQueryer struct {
+	cloud.CloudQueryer
+	*jujutesting.CallMocker
+}
+
+func (f *fakeCloudQueryer) ParseCloudMetadataFile(path string) (map[string]cloudfile.Cloud, error) {
+	results := f.MethodCall(f, "ParseCloudMetadataFile", path)
+	return results[0].(map[string]cloudfile.Cloud), typeAssertError(results[1])
+}
+
+func (f *fakeCloudQueryer) PublicCloudMetadata(searchPaths ...string) (result map[string]cloudfile.Cloud, fallbackUsed bool, _ error) {
+	results := f.MethodCall(f, "PublicCloudMetadata", searchPaths)
```

Comment:
Well, so like, if you change the name of PublicCloudMetadata, you could forget or typo the change to the string here or in the test, and you'd get weird behavior from your test (likely your call would return nil or even possibly panic).  Yes, it likely wouldn't pass, but it would be hard to debug.  And while this method is wrapped in something type safe, the .Call() and .Returns() method in the tests aren't.  If you change the signature of PublicCloudMetadata, the tests would still compile, even though they're returning the wrong types.  Sure, the tests wouldn't pass, but it might be really hard to figure out if they're returning int64(5) vs int(5) for example.

I just don't see that the saving on typing is worth the loss of type safety and the possibility of simple typos causing lost developer time.